test/metabase/db/connection_test.clj

```diff
@@ -138,3 +141,24 @@
     (with-open [conn (.getConnection mdb.connection/*application-db*)]
       (is (= java.sql.Connection/TRANSACTION_READ_COMMITTED
              (.getTransactionIsolation conn))))))
+
+(deftest rollback-error-handling
+  (testing "rollback error handling"
+    (let [real-conn (.getConnection ^DataSource mdb.connection/*application-db*)
+          mock-conn (reify Connection
+                      (rollback [_ _savepoint]
+                        (throw (ex-info "Rollback error" {})))
+                      (setAutoCommit [_ auto?]
+                        (.setAutoCommit real-conn auto?))
+                      (getAutoCommit [_]
+                        (.getAutoCommit real-conn))
+                      (setSavepoint [_]
+                        (.setSavepoint real-conn))
+                      (commit [_]
+                        (.commit real-conn)))]
```

Comment:
I think you can get this test to work without acquiring a real connection? Just mock all the other methods as well
```suggestion
    (let [mock-conn (reify Connection
                      (rollback [_ _savepoint]
                        (throw (ex-info "Rollback error" {})))
                      (setAutoCommit [_ _auto?])
                      (getAutoCommit [_] true)
                      (setSavepoint [_])
                      (commit [_]))]
```

If you do want to use a real connection, probably better to use `with-open` so that it gets automatically closed.